### PR TITLE
AB#11693

### PR DESCRIPTION
**Pull request by Matthis-M-ReliefApps**
# Description

Until now, the way to prefill a form with a record was the following:
1. In grid settings, add an action button with "Go to next step" and "prefill form" option
2. Add the form you want to prefill after the grid in a workflow
3. Select the record(s) you want to prefill the form with
4. Click on the "go to next step" action button
Result => The user navigate to the next step of the workflow, the form, with prefilled fields

This PR aims to make the process much simpler and intuitive for the user, by changing the flow to the following:
1. In grid settings, add an action button with "prefill form" option, and select the form you want to prefill
2. Select the record(s) you want to prefill the form with
3. Click on the "prefill form" action button
Result => A model opens containing the form, prefilled with the selected record(s)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (refactor or addition to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] In a grid, add an action button with aforementioned settings. Select a record, and use the button. The form modal should open with its fields prefilled
- [ ] In a grid, add an action button with aforementioned settings. Select multiple records, and use the button. The form modal should open with its fields prefilled and the possibility to fill multiple forms this way.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: projects/safe/src/lib/components/widgets/grid/grid.component.ts [hunk 1]
```diff
@@ -1032,7 +1038,18 @@ export class SafeGridComponent implements OnInit, OnChanges, OnDestroy {
...
-        this.workflowService.storeRecords(records);
+
+
+        // Open a modal containing the prefilled form
+        this.dialog.open(SafeFormModalComponent, {
+          data: {
+            recordId: selectedRecords.map(x => x.id),
+            template: options.prefillTargetForm,
+            locale: 'en',
+            prefillRecords: records,
```
Comment (AntoineRelief) [inline projects/safe/src/lib/components/widgets/grid/grid.component.ts:1052]: @Matthis-M-ReliefApps I'd have built a similar logic, moving the logic of the nextStep to the SafeFormModal when passing the data.

nevertheless I don't think you need to pass recordId there, as we want to create a blank record, containing some data

File: projects/safe/src/lib/components/form/form.component.ts [hunk 2]
```diff
@@ -103,43 +95,13 @@ export class SafeFormComponent implements OnInit, OnDestroy, AfterViewInit {
...
-          if (resourcesField && resourcesField.resource === records[0].form?.resource?.id) {
-            cachedData[resourcesField.name] = records.map(x => x.id);
-          } else {
-            this.snackBar.openSnackBar(NOTIFICATIONS.recordDoesNotMatch, { error: true });
-          }
-        }
-      });
-    }
-
-    if (this.form.uniqueRecord && this.form.uniqueRecord.data) {
```
Comment (AntoineRelief) [inline projects/safe/src/lib/components/form/form.component.ts:129]: @Matthis-M-ReliefApps not something you'd seen, but this is important to keep
In fact, there's a use case where we can tell that a form has an unique record per user / per role etc

**Comment by AntoineRelief** [comment]
@Matthis-M-ReliefApps almost all good for this one, I only changed one thing:
in fact the prefill form should be one of the forms related to the current form ( so for ex, if current form refers to another form, then it's related form )
so instead of templates, we would use relatedForms

File: projects/safe/src/lib/components/form/form.component.ts (hunk 2 above)
Comment (Matthis-M-ReliefApps) [reply projects/safe/src/lib/components/form/form.component.ts:129]: Ahah I was sure I had deleted at least one important thing while doing the "cleaning"

**Pull request merged**